### PR TITLE
Remove the DPCTLQueueMgr_* and device_context API from dpctl

### LINKED ISSUE #11: Thread-safety of `device_context`
Just to record the issues raised by @hameerabbasi and I in the numba dev call:

* If two threads use `with device_context` in parallel, do they interfere with each other?
* If two async tasks use `with_device_context` in parallel, do they interfere with each other? (xref PEP 567)
* Can a child thread reuse the queue of a parent thread?

### LINKED ISSUE #410: Add tests for asyncio safety, and Python parallelism safety
Basically add tests to cover concerns raised in #11 

Testing of asyncio safety will require adding test-time dependency on `pytest-asyncio`. The test file may be as follows:

```python
import dpctl
import dpctl.memory as dpmem
import random
import asyncio

async def _task1():
    q = dpctl.SyclQueue("opencl:gpu")
    abc = b"abcdefghijklmnopqrstuvwxyz"
    m = dpmem.MemoryUSMShared(len(abc))
    with dpctl.device_context(q) as lq:
        for _ in range(12):
            cd = dpctl.get_current_queue().sycl_device
            assert cd.backend == q.sycl_device.backend
            m.copy_from_host(abc)
            await asyncio.sleep(0.1 * random.random())


async def _task2():
    q = dpctl.SyclQueue("level_zero:gpu")
    m = dpmem.MemoryUSMShared(10)
    host_data = b"\x00" * 10
    with dpctl.device_context(q) as lq:
        for _ in range(12):
            cd = dpctl.get_current_queue().sycl_device
            assert cd.backend == q.sycl_device.backend
            m.copy_from_host(host_data)
            await asyncio.sleep(0.1 * random.random())


async def test_asyncio_safety():
    j1 = asyncio.create_task(_task1())
    j2 = asyncio.create_task(_task2())
    await j1
    await j2

    print("done")



if __name__ == '__main__':
    asyncio.run(test_asyncio_safety())
```

The test for multi-processing may look as follows, but to turn it into a test, assertions must be added:

```python
import dpctl
import dpctl.memory as dpmem
import multiprocessing as mp


def compute_works(i):
    d = dpctl.SyclDevice("cpu")
    sd = d.create_sub_devices(partition=(4, 4, 4))
    ctx = dpctl.SyclContext(sd)
    q = dpctl.SyclQueue(ctx, sd[i], property="in_order")
    mem = dpmem.MemoryUSMShared(32, queue=q)
    host = (b" ") * 32
    mem.copy_from_host(host)
    print((i, "->", mem.copy_to_host()))
    print("", flush=True)
    # dpctl.set_global_queue(q)


def rt_warnings(i):
    cq = dpctl.get_current_queue()
    with dpctl.device_context("cpu") as q:
        print(
            "Hello from {}, using {} with {} EUs".format(
                i, q.sycl_device.name, q.sycl_device.max_compute_units
            )
        )
        dpctl.set_global_queue(q)
    return i


if __name__ == "__main__":
    with mp.Pool(3) as p:
        p.map(rt_warnings, [0, 1, 2])

    print("Execution of rt_warnings finished")
    print("--" * 30)

    with mp.Pool(3) as p:
        p.map(compute_works, [0, 1, 2])

    print("Execution of compute_works finished")
```

### LINKED ISSUE #1575: Removal of libsyclinterface DPCTLQueueMgr_* functions
The `dpctl_sycl_queue_manager` is a legacy of an old design in DPEP stack where we were hoping to use a single Python context manager to define the execution queue globally. DPEP software stack now uses the array API compute follows data programming model, as such the following functions are redundant and should be removed:

> 
> DPCTL_API __dpctl_give [DPCTLSyclQueueRef](https://intelpython.github.io/dpctl/latest/docfiles/libsyclinterface/global.html#doxid-df-d57-dpctl-sycl-types-8h-1af23137a8bf20ff63b10db9f7f987799a)
> [DPCTLQueueMgr_GetCurrentQueue](https://intelpython.github.io/dpctl/latest/docfiles/libsyclinterface/group_QueueManager.html#doxid-d0-dde-group-queue-manager-1gaa1e5044153f900fff9b779b30f3b0538)(void);
> 
> DPCTL_API size_t
> [DPCTLQueueMgr_GetQueueStackSize](https://intelpython.github.io/dpctl/latest/docfiles/libsyclinterface/group_QueueManager.html#doxid-d0-dde-group-queue-manager-1gac4dacf08c61ae001907660867185ea7a)(void);
> 
> DPCTL_API bool
> [DPCTLQueueMgr_GlobalQueueIsCurrent](https://intelpython.github.io/dpctl/latest/docfiles/libsyclinterface/group_QueueManager.html#doxid-d0-dde-group-queue-manager-1ga38b8e0f4ea42601d27eda0d5413c5164)(void);
> 
> DPCTL_API bool
> [DPCTLQueueMgr_IsCurrentQueue](https://intelpython.github.io/dpctl/latest/docfiles/libsyclinterface/group_QueueManager.html#doxid-d0-dde-group-queue-manager-1ga77d39c59561d3f427a1dde1ce81652db)(__dpctl_keep const [DPCTLSyclQueueRef](https://intelpython.github.io/dpctl/latest/docfiles/libsyclinterface/global.html#doxid-df-d57-dpctl-sycl-types-8h-1af23137a8bf20ff63b10db9f7f987799a) QRef);
> 
> DPCTL_API void
> [DPCTLQueueMgr_PopQueue](https://intelpython.github.io/dpctl/latest/docfiles/libsyclinterface/group_QueueManager.html#doxid-d0-dde-group-queue-manager-1gac5fa7690c281328a070f2409b548df73)(void);
> 
> DPCTL_API void
> [DPCTLQueueMgr_PushQueue](https://intelpython.github.io/dpctl/latest/docfiles/libsyclinterface/group_QueueManager.html#doxid-d0-dde-group-queue-manager-1gac9c2b3d22c3bfb0797c0b35551d331b7)(__dpctl_keep const [DPCTLSyclQueueRef](https://intelpython.github.io/dpctl/latest/docfiles/libsyclinterface/global.html#doxid-df-d57-dpctl-sycl-types-8h-1af23137a8bf20ff63b10db9f7f987799a) QRef);
> 
> DPCTL_API void
> [DPCTLQueueMgr_SetGlobalQueue](https://intelpython.github.io/dpctl/latest/docfiles/libsyclinterface/group_QueueManager.html#doxid-d0-dde-group-queue-manager-1gaf767ec0189383390b6bc63835eef1543)(__dpctl_keep const [DPCTLSyclQueueRef](https://intelpython.github.io/dpctl/latest/docfiles/libsyclinterface/global.html#doxid-df-d57-dpctl-sycl-types-8h-1af23137a8bf20ff63b10db9f7f987799a) QRef);
> 
> 

On dpctl's side I believe the whole dpctl._sycl_queue_manager.pyx can be removed.

### PR DESCRIPTION
**Pull request by diptorupd**
- [X] Have you provided a meaningful PR description?

The `DPCTLQueueMgr_*` implemented a global stack in libsyclinterface to keep track of a "current queue". The stack was controlled from dpctl's Python API using `dpctl.device_context`. Both these features are legacy design choices that have been abandoned and no longer are required. The PR removes the DPCTLQueueMgr_*` interface, `dpctl.device_context` and associated functions, removes and updates test cases as needed.

- [X] Have you added a test, reproducer or referred to an issue with a reproducer?

Existing tests were updated.
- [X] Have you tested your changes locally for CPU and GPU devices?
- [X] Have you made sure that new changes do not introduce compiler warnings?
- [ ] Have you checked performance impact of proposed changes?
- [ ] If this PR is a work in progress, are you opening the PR as a draft?

Fixes #11
Fixes #410
Fixes #1575

### CODE REVIEW

File: libsyclinterface/tests/test_sycl_usm_interface.cpp [hunk 1]
```diff
@@ -95,67 +94,68 @@ struct TestDPCTLSyclUSMInterface : public ::testing::Test
...
     EXPECT_TRUE(bool(Ptr));
     common_test_body(nbytes, Ptr, Q, DPCTLSyclUSMType::DPCTL_USM_SHARED);
     DPCTLfree_with_queue(Ptr, Q);
-    DPCTLQueue_Delete(Q);
 }
 
 TEST_F(TestDPCTLSyclUSMInterface, AlignedAllocDevice)
 {
-    auto Q = DPCTLQueueMgr_GetCurrentQueue();
+    sycl::queue q;
```
Comment (oleksandr-pavlyk) [inline libsyclinterface/tests/test_sycl_usm_interface.cpp:145]: I think we should use `DPCTLQueue_CreateForDevice` and create device from device selector

**Review by oleksandr-pavlyk — CHANGES_REQUESTED** [review]

File: libsyclinterface/tests/test_sycl_usm_interface.cpp (hunk 1 above)
Comment (oleksandr-pavlyk) [reply libsyclinterface/tests/test_sycl_usm_interface.cpp]: I pushed changes to implement my suggestion while rebasing this branch to resolve conflicts.

**Review by oleksandr-pavlyk — COMMENTED** [review]
This looks good to me know, but it would be best if `dpnp` can resolve its dependency on `DPCTLQueueMgr_*` before merging this. 

One option is to vendor implementation of `DPCTLQueueMgr_GetCurrentQueue` without implementing any code to modify the stack.

**Comment by diptorupd** [comment]
> This looks good to me know, but it would be best if `dpnp` can resolve its dependency on `DPCTLQueueMgr_*` before merging this.
> 
> One option is to vendor implementation of `DPCTLQueueMgr_GetCurrentQueue` without implementing any code to modify the stack.

It is okay to wait for dpnp. Are there any blockers?

**Comment by oleksandr-pavlyk** [comment]
`dpnp` pr IntelPython/dpnp#1735 has been merged. No blockers to merge this one.

**Review by oleksandr-pavlyk — APPROVED** [review]

**Pull request merged**